File: src/main/java/toby/command/commands/KickCommand.java
package toby.command.commands;

import toby.command.CommandContext;
import toby.command.ICommand;

import java.util.List;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.TextChannel;

public class KickCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) {
        final TextChannel channel = ctx.getChannel();
        final Message message = ctx.getMessage();
        final Member member = ctx.getMember();
        final List<String> args = ctx.getArgs();

        if (message.getMentionedMembers().isEmpty()) {
            channel.sendMessage("You must mention 1 or more Users to shoot").queue();
            return;
        }

        message.getMentionedMembers().forEach(target -> {

        if (!member.canInteract(target) || !member.hasPermission(Permission.KICK_MEMBERS)) {
            channel.sendMessage(String.format("You can't shoot %s", target)).queue();
            return;
        }

        final Member selfMember = ctx.getSelfMember();

        if (!selfMember.canInteract(target) || !selfMember.hasPermission(Permission.KICK_MEMBERS)) {
            channel.sendMessage(String.format("I'm not allowed to shoot %s", target)).queue();
            return;
        }

        ctx.getGuild()
                .kick(target)
                .reason("because you told me to.")
                .queue(
                        (__) -> channel.sendMessage("Shot hit the mark... something about fortnite?").queue(),
                        (error) -> channel.sendMessageFormat("Could not shoot %s", error.getMessage()).queue()
                );
        });
    }

    @Override
    public String getName() {
        return "kick";
    }

    @Override
    public String getHelp() {
        return "Kick a member off the server.\n" +
                "Usage: `!kick <@user>`";
    }
}

File: discord-bot/src/main/kotlin/bot/toby/command/commands/moderation/KickCommand.kt
package bot.toby.command.commands.moderation

import bot.toby.command.PermissionValidator
import core.command.Command.Companion.replyAndDelete
import core.command.CommandContext
import database.dto.user.UserDto
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class KickCommand @Autowired constructor(
    private val permissionValidator: PermissionValidator,
) : ModerationCommand {
    companion object {
        private const val USERS = "users"
    }

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int) {
        val event = ctx.event

        val guild = event.guild!!
        val member = ctx.member ?: return
        val botMember = guild.selfMember

        val mentionedMembers = event.getOption(USERS)
            ?.mentions
            ?.members

        if (mentionedMembers.isNullOrEmpty()) {
            event.hook.replyAndDelete("You must mention 1 or more Users to kick.", deleteDelay)
            return
        }

        mentionedMembers.forEach { target ->
            val actorOk = permissionValidator.actorMayActOn(
                event, member, target, Permission.KICK_MEMBERS, "kick", deleteDelay
            )
            if (!actorOk) return@forEach
            val botOk = permissionValidator.botMayAct(
                event, botMember, target, Permission.KICK_MEMBERS, "kick", deleteDelay,
                requireCanInteract = true,
            )
            if (!botOk) return@forEach
            guild.kick(target).reason("because you told me to.").queue(
                {
                    event.hook.replyAndDelete(
                        "Shot hit the mark... something about fortnite?",
                        deleteDelay,
                    )
                },
                { error ->
                    event.hook.replyAndDelete("Could not kick: ${error.message}", deleteDelay)
                }
            )
        }
    }

    override val name: String
        get() = "kick"

    override val description: String
        get() = "Kick a member off the server."

    override val optionData: List<OptionData>
        get() = listOf(
            OptionData(OptionType.STRING, USERS, "User(s) to kick", true)
        )
}
